### PR TITLE
bitswap: protocol extension to handle cids

### PR DESCRIPTION
**Pull request by whyrusleeping**
This change adds the /ipfs/bitswap/1.1.0 protocol. The new protocol
adds a 'payload' field to the protobuf message and deprecates the
existing 'blocks' field. The 'payload' field is an array of pairs of cid
prefixes and block data. The cid prefixes are used to ensure the correct
codecs and hash functions are used to handle the block on the receiving
end.

cc @diasdavid @Kubuxu 

License: MIT
Signed-off-by: Jeromy <why@ipfs.io>

### CODE REVIEW

File: exchange/bitswap/message/message.go [hunk 1]
```diff
@@ -78,10 +80,36 @@ func newMessageFromProto(pbm pb.Message) (BitSwapMessage, error) {
 		}
 		m.addEntry(c, int(e.GetPriority()), e.GetCancel())
 	}
+
+	// deprecated
 	for _, d := range pbm.GetBlocks() {
+		fmt.Println("got old style block")
```
Comment (Kubuxu) [inline exchange/bitswap/message/message.go]: This isn't needed.

File: exchange/bitswap/message/message.go [hunk 2]
```diff
@@ -78,10 +80,36 @@ func newMessageFromProto(pbm pb.Message) (BitSwapMessage, error) {
...
+		// CIDv0, sha256, protobuf only
 		b := blocks.NewBlock(d)
 		m.AddBlock(b)
 	}
+	//
+
+	for _, b := range pbm.GetPayload() {
+		fmt.Println("got new style block")
```
Comment (Kubuxu) [inline exchange/bitswap/message/message.go]: This is also leftover debug logging.

File: exchange/bitswap/message/message.go [hunk 3]
```diff
@@ -78,10 +80,36 @@ func newMessageFromProto(pbm pb.Message) (BitSwapMessage, error) {
...
 	}
+
+	// deprecated
 	for _, d := range pbm.GetBlocks() {
+		fmt.Println("got old style block")
+		// CIDv0, sha256, protobuf only
 		b := blocks.NewBlock(d)
 		m.AddBlock(b)
 	}
+	//
```
Comment (Kubuxu) [inline exchange/bitswap/message/message.go]: This isn't needed either.

File: exchange/bitswap/message/message.go [hunk 4]
```diff
@@ -78,10 +80,36 @@ func newMessageFromProto(pbm pb.Message) (BitSwapMessage, error) {
 		}
 		m.addEntry(c, int(e.GetPriority()), e.GetCancel())
 	}
+
+	// deprecated
 	for _, d := range pbm.GetBlocks() {
+		fmt.Println("got old style block")
+		// CIDv0, sha256, protobuf only
 		b := blocks.NewBlock(d)
 		m.AddBlock(b)
```
Comment (Kubuxu) [inline exchange/bitswap/message/message.go]: What in the case that both v0 and v1 fields are used? Just allow it? Log it?

File: exchange/bitswap/message/message.go [hunk 5]
```diff
@@ -169,10 +197,41 @@ func (m *impl) ToProto() *pb.Message {
...
+			Data:   b.RawData(),
+			Prefix: b.Cid().Prefix().Bytes(),
+		}
+		pbm.Payload = append(pbm.Payload, blk)
+	}
+	return pbm
+}
+
+func (m *impl) ToNetV0(w io.Writer) error {
+	fmt.Println("V0")
```
Comment (Kubuxu) [inline exchange/bitswap/message/message.go]: Again debug logging.

File: exchange/bitswap/message/message.go (hunk 1 above)
Comment (whyrusleeping) [inline exchange/bitswap/message/message.go]: oh right, those debug logs...

**Comment by whyrusleeping** [comment]
I have two more PR's ready to go (already written) once this one is in. So lets get review on this ASAP

**Comment by Kubuxu** [comment]
LGTM

**Pull request merged**